pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -148,56 +146,136 @@ func NewFlowExporter(connStore *connections.ConntrackConnectionStore, records *f
+				// connectivity, we reset the connection to IPFIX collector and retry
+				// in the next export cycle to reinitialize the connection and send flow records.
+				exp.process.CloseConnToCollector()
+				exp.process = nil
+				expireTimer.Reset(exp.expirePriorityQueue.ActiveFlowTimeout)
+				continue
+			}
+
+			// Get the new expiry and reset the timer.
+			expireTimer.Reset(exp.GetExpiryFromExpirePriorityQueue())
 		}
 	}
-	// Send flow records to IPFIX collector.
-	err := exp.sendFlowRecords()
-	if err != nil {
-		klog.Errorf("Error when sending flow records: %v", err)
-		// If there is an error when sending flow records because of intermittent connectivity, we reset the connection
-		// to IPFIX collector and retry in the next export cycle to reinitialize the connection and send flow records.
-		exp.process.CloseConnToCollector()
-		exp.process = nil
-		return
+}
+
+// GetExpiryFromExpirePriorityQueue returns the earliest timestamp (active expiry
+// or idle expiry) from expire priority queue.
+func (exp *flowExporter) GetExpiryFromExpirePriorityQueue() time.Duration {
+	exp.expirePriorityQueue.Mutex.Lock()
+	defer exp.expirePriorityQueue.Mutex.Unlock()
+
+	currTime := time.Now()
+	if exp.expirePriorityQueue.Len() > 0 {
+		// Get the minExpireTime of the top item in ExpirePriorityQueue.
+		expiryDuration := MinExpiryTime + exp.expirePriorityQueue.MinExpireTime(0).Sub(currTime)
+		if expiryDuration < 0 {
+			return MinExpiryTime
+		}
+		return expiryDuration
+	}
+	if exp.expirePriorityQueue.ActiveFlowTimeout < exp.expirePriorityQueue.IdleFlowTimeout {
+		return exp.expirePriorityQueue.ActiveFlowTimeout
+	}
+	return exp.expirePriorityQueue.IdleFlowTimeout
+}
+
+func (exp *flowExporter) sendFlowRecords() error {
+	exp.expirePriorityQueue.Mutex.Lock()
+	defer exp.expirePriorityQueue.Mutex.Unlock()
+
+	if exp.expirePriorityQueue.Len() == 0 {
+		return nil
+	}
+	currTime := time.Now()
+	for exp.expirePriorityQueue.Len() > 0 {
+		topItem := exp.expirePriorityQueue.Peek()
+		// If the top item is not timeout, then we do not have to check the following items.
+		if topItem.ActiveExpireTime.After(currTime) && topItem.IdleExpireTime.After(currTime) {
+			break
+		}
+		pqItem := heap.Pop(exp.expirePriorityQueue).(*flowexporter.ItemToExpire)
+		if !pqItem.IsDenyConn {
+			// The pqItem contains a conntrack connection.
+			exp.sendItemToExpire(pqItem, false)
+			if flowexporter.IsConnectionDying(&pqItem.Conn) {
+				// If the connection is in dying state or connection is not in conntrack
+				// table, we set the DyingAndDoneExport flag to do the deletion later.
+				pqItem.DyingAndDoneExport = true
+			} else {
+				// If there is no increase in stats, we set IsActive flag to false,
+				// and do not push back into the PQ.
+				if pqItem.IdleExpireTime.Before(currTime) && (pqItem.Conn.OriginalPackets <= pqItem.PrevPackets) && (pqItem.Conn.ReversePackets <= pqItem.PrevReversePackets) {
```

Comment:
do we have to check stats again as the idle expiry is already triggered?

Reply:
we can set IsActive = false once an item is idle time out. will do.